File: longest-common-prefix/longest-common-prefix.js
/**
 * @param {string[]} strs
 * @return {string}
 */
var longestCommonPrefix = function(strs) {
    if (strs.length == 1) return strs[0]
    
    let result = ""

    strs.sort() // sort alphabetically, so we can just compare the first and last element only because if it's sorted the middle words will be automatically same as first and last

    for (let i = 0; i < strs[0].length; i++) {
        if (!strs[0].length) return ""

        // check every char for first and last index
        if (strs[0][i] == strs[strs.length-1][i]) { 
            result += strs[0][i]
        } 
        // stop compare whenever the char is not match and return result   immediately
        else { 
        break
        }
    }

    return result
};

File: longest-common-prefix/longest-common-prefix.go
package main

import (
	"fmt"
	"sort"
	// "strings"
)

func main() {

	arrStr := []string{"adrian", "adroid", "adrimas"}
	fmt.Println(longestCommonPrefix(arrStr))
}

func longestCommonPrefix(strs []string) string {
	var longestPrefix string = ""

	if len(strs) > 0 {
		sort.Strings(strs) // sor alphabetically so we can just compare frist and last string
		first := string(strs[0])
		last := string(strs[len(strs)-1])

		fmt.Println(strs)
		for i := 0; i < len(first); i++ {
			if string(last[i]) == string(first[i]) {
				longestPrefix += string(last[i])
			} else {
				return longestPrefix
			}
		}
	}
	return longestPrefix

}
